core-data/queried-data/selectors.js

```diff
@@ -0,0 +1,89 @@
+/**
+ * External dependencies
+ */
+import { get } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import getQueryParts from './get-query-parts';
+
+/**
+ * Utility function returning the current value of `requestingPageByPerPage`
+ * for the given state and optional query. Returns undefined if no requests
+ * have been initiated for the given query page, otherwise a boolean value
+ * representing whether a request is in progress.
+ *
+ * @param {Object}  state State object.
+ * @param {?Object} query Optional query.
+ *
+ * @return {?boolean} `requestingPageByPerPage` value for query.
+ */
+function getRequestingPageByPerPage( state, query ) {
+	const { stableKey, page, perPage } = getQueryParts( query );
+
+	return get( state, [
+		'queries',
+		stableKey,
+		'requestingPageByPerPage',
+		perPage,
+		page,
+	] );
+}
+
+/**
+ * Returns items for a given query, or null if the items are not known.
+ *
+ * @param {Object}  state State object.
+ * @param {?Object} query Optional query.
+ *
+ * @return {?Array} Query items.
+ */
+export function getQueriedItems( state, query ) {
```

Comment:
One issue with memoization is the same as described at https://github.com/WordPress/gutenberg/pull/6360#issuecomment-383682602 . For example, consider:

```js
export default withSelect( ( select ) => {
	const { getCategories } = select( 'core' );

	return {
		categories: getCategories( { order: 'desc' } ),
	};
} )( CategoriesBlock );
```

In this case, we'd encounter a memory leak in that each object reference would be strictly non-equal.

The `getQueryParts` function gets around this by using the `withWeakMapCache` helper, which is designed to alleviate this by caching only as long as the object reference is in scope, otherwise allowing garbage collection to occur. Unfortunately that's only effective where we have a single object argument, not two.

We _could_ perhaps keep a local WeakMap cache in scope, and do memoization manually.

Reply:
Actually, I'm getting my own caching libraries confused 😄 In `memize`, this would be an issue. `rememo` is much smarter though, particularly for functions like this receiving object arguments ([source](https://github.com/aduth/rememo/blob/d40eaa1af3d344aefc7ff03bae8ebdb6f1500ce5/es/rememo.js#L117-L170)). I think it would work well here.

Reply:
Can we have a specific memoization here where we compare the first argument using strict equality (state) and the second one using deep equality (using `EquivalentKeyMap ` maybe)?